Bottom: 40162b03f0d0298c368821e998f6b7b54781a942
Top:    ed212559043bacac232c481a125c8cfff34c5b04
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-04-05 22:39:50 +0100

GTK HID: Use clipping region when drawing DRC violation previews

Should speed up rendering quite a bit for non-trivial boards.


---

diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index abf0b3c..c27f857 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -58,71 +58,17 @@ int ghid_flip_x = 0, ghid_flip_y = 0;
 
 /* ------------------------------------------------------------ */
 
-/* Px converts view->pcb, Vx converts pcb->view */
-      
-static inline int 
-Vx (int x)
-{     
-  int rv;
-  if (ghid_flip_x) 
-    rv = (PCB->MaxWidth - x - gport->view_x0) / gport->zoom + 0.5;
-  else
-    rv = (x - gport->view_x0) / gport->zoom + 0.5;
-  return rv;
-}       
-      
 static inline int 
 Vx2 (int x)
 {     
   return (x - gport->view_x0) / gport->zoom + 0.5;
 }       
-      
-static inline int
-Vy (int y)
-{         
-  int rv;
-  if (ghid_flip_y)
-    rv = (PCB->MaxHeight - y - gport->view_y0) / gport->zoom + 0.5;
-  else
-    rv = (y - gport->view_y0) / gport->zoom + 0.5;
-  return rv;
-}     
         
 static inline int 
 Vy2 (int y)
 {     
   return (y - gport->view_y0) / gport->zoom + 0.5;
 }       
-      
-static inline int
-Vz (int z)
-{           
-  return z / gport->zoom + 0.5;
-}         
-                
-static inline int
-Px (int x)
-{  
-  int rv = x * gport->zoom + gport->view_x0;
-  if (ghid_flip_x)
-    rv = PCB->MaxWidth - (x * gport->zoom + gport->view_x0);
-  return  rv;
-}  
-
-static inline int
-Py (int y)
-{  
-  int rv = y * gport->zoom + gport->view_y0;
-  if (ghid_flip_y)
-    rv = PCB->MaxHeight - (y * gport->zoom + gport->view_y0);
-  return  rv;
-}  
-
-static inline int  
-Pz (int z)
-{
-  return (z * gport->zoom);
-}
 
 /* ------------------------------------------------------------ */
 
diff --git a/src/hid/gtk/gui-render-pixmap.c b/src/hid/gtk/gui-render-pixmap.c
index 525bde6..b24e6c9 100644
--- a/src/hid/gtk/gui-render-pixmap.c
+++ b/src/hid/gtk/gui-render-pixmap.c
@@ -49,6 +49,7 @@ ghid_render_pixmap (int cx, int cy, double zoom, int width, int height, int dept
   int save_left, save_top;
   int save_width, save_height;
   int save_view_width, save_view_height;
+  BoxType region;
 
   save_drawable = gport->drawable;
   save_zoom = gport->zoom;
@@ -79,7 +80,11 @@ ghid_render_pixmap (int cx, int cy, double zoom, int width, int height, int dept
   gdk_draw_rectangle (pixmap, gport->bg_gc, TRUE, 0, 0, width, height);
 
   /* call the drawing routine */
-  hid_expose_callback (&ghid_hid, NULL, NULL);
+  region.X1 = MIN(Px(0), Px(gport->width + 1));
+  region.Y1 = MIN(Py(0), Py(gport->height + 1));
+  region.X2 = MAX(Px(0), Px(gport->width + 1));
+  region.Y2 = MAX(Py(0), Py(gport->height + 1));
+  hid_expose_callback (&ghid_hid, &region, NULL);
 
   gport->drawable = save_drawable;
   gport->zoom = save_zoom;
diff --git a/src/hid/gtk/gui.h b/src/hid/gtk/gui.h
index c94c2a3..1ed9082 100644
--- a/src/hid/gtk/gui.h
+++ b/src/hid/gtk/gui.h
@@ -532,11 +532,63 @@ void ghid_invalidate_all ();
 void ghid_get_coords (const char *msg, int *x, int *y);
 gint PCBChanged (int argc, char **argv, int x, int y);
 
-
-
-
 extern GdkPixmap *XC_hand_source, *XC_hand_mask;
 extern GdkPixmap *XC_lock_source, *XC_lock_mask;
 extern GdkPixmap *XC_clock_source, *XC_clock_mask;
 
+
+/* Coordinate conversions */
+/* Px converts view->pcb, Vx converts pcb->view */
+static inline int
+Vx (int x)
+{
+  int rv;
+  if (ghid_flip_x)
+    rv = (PCB->MaxWidth - x - gport->view_x0) / gport->zoom + 0.5;
+  else
+    rv = (x - gport->view_x0) / gport->zoom + 0.5;
+  return rv;
+}
+
+static inline int
+Vy (int y)
+{
+  int rv;
+  if (ghid_flip_y)
+    rv = (PCB->MaxHeight - y - gport->view_y0) / gport->zoom + 0.5;
+  else
+    rv = (y - gport->view_y0) / gport->zoom + 0.5;
+  return rv;
+}
+
+static inline int
+Vz (int z)
+{
+  return z / gport->zoom + 0.5;
+}
+
+static inline int
+Px (int x)
+{
+  int rv = x * gport->zoom + gport->view_x0;
+  if (ghid_flip_x)
+    rv = PCB->MaxWidth - (x * gport->zoom + gport->view_x0);
+  return  rv;
+}
+
+static inline int
+Py (int y)
+{
+  int rv = y * gport->zoom + gport->view_y0;
+  if (ghid_flip_y)
+    rv = PCB->MaxHeight - (y * gport->zoom + gport->view_y0);
+  return  rv;
+}
+
+static inline int
+Pz (int z)
+{
+  return (z * gport->zoom);
+}
+
 #endif /* __GHID_INCLUDED__  */
